src/Init/Data/Cast.lean

```diff
@@ -0,0 +1,101 @@
+/-
+Copyright (c) 2014 Mario Carneiro. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Mario Carneiro, Gabriel Ebner
+-/
+prelude
+import Init.Data.Nat.Basic
+import Init.Data.Int.Basic
+
+/-!
+# `NatCast` and `IntCast`
+
+We introduce the typeclass `NatCast R` for a type `R` with a "canonical homomorphism" `Nat → R`.
+The typeclass carries the data of the function, but no required axioms.
+
+This typeclass was introduced to a uniform simp normal form for such morphisms.
+
+Without such a typeclass, we would have coercions such as `Int.ofNat`,
+but also later in Mathlib the generic coercion from `Nat` into any semiring (including `Int`),
+and we would need to use `simp` to move between them.
+However `simp` lemmas expressed using a non-normal form on the LHS would then not fire.
+
+In order to avoid this, we introduce this typeclass as early as possible,
+ensure it is a higher priority coercion that built-in coercions such as `Int.ofNat`,
```

Comment:
Now that this is being upstreamed, what about deleting `Int.ofNat` and using only this one? That would simplify things a bit.

Reply:
We can't delete `Int.ofNat`, it's one of the constructors of `Int`

Reply:
Hah, that's right! Thanks :-) Perhaps it can stop being a coercion, then, once we have this merged?

Reply:
(my concern here is keeping things as simple as possible - if this would be too simple, no problem, but I think it's worth asking)

Reply:
Yes, I'm hoping we can do that. But it will happen after this PR. I've added it to the Std todo list.